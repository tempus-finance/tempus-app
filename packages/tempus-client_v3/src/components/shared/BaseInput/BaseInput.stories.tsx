import { ComponentStory } from '@storybook/react';
import { FC, useCallback, useState } from 'react';
import BaseInput, { BaseInputProps } from './BaseInput';
import Typography from '../Typography';

export default {
  title: 'BaseInput',
  component: BaseInput,
  argTypes: {},
};

const style = {
  background: 'rgba(0, 0, 0, 0.1)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '10px',
};

const Wrapper: FC<BaseInputProps> = props => {
  const [value, setValue] = useState<string>(props.value ?? '');
  const [debouncedValue, setDebouncedValue] = useState<string>('');
  const onChange = useCallback((val: string) => setValue(val), []);
  const onDebounceChange = useCallback((val: string) => setDebouncedValue(val), []);
  return (
    <div>
      <BaseInput {...props} value={value} onChange={onChange} onDebounceChange={onDebounceChange} />
      <hr />
      <Typography variant="body-secondary">onChange: {value}</Typography>
      <Typography variant="body-secondary">onDebounceChange: {debouncedValue}</Typography>
    </div>
  );
};

const Template: ComponentStory<typeof BaseInput> = args => (
  <div style={style}>
    <Wrapper {...args} />
  </div>
);

export const BaseInputNormal = Template.bind({});
BaseInputNormal.args = {
  value: 'value',
  placeholder: 'placeholder',
  pattern: '',
  disabled: false,
  debounce: false,
};

export const BaseInputDisabled = Template.bind({});
BaseInputDisabled.args = {
  value: 'value',
  placeholder: 'placeholder',
  pattern: '',
  disabled: true,
  debounce: false,
};

export const BaseInputLetterOnly = Template.bind({});
BaseInputLetterOnly.args = {
  value: 'value',
  placeholder: 'placeholder',
  pattern: '[a-zA-Z]*',
  disabled: false,
  debounce: false,
};

export const BaseInputDebounce = Template.bind({});
BaseInputDebounce.args = {
  value: 'value',
  placeholder: 'placeholder',
  pattern: '',
  disabled: false,
  debounce: true,
};
