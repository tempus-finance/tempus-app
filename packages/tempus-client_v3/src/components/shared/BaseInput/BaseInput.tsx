import { ChangeEvent, FocusEvent, FC, useCallback, useMemo, useRef } from 'react';

export interface BaseInputProps {
  id?: string;
  value?: string;
  placeholder?: string;
  pattern?: string;
  disabled?: boolean;
  debounce?: boolean | number;
  onChange?: (value: string) => void;
  onDebounceChange?: (value: string) => void;
  onFocus?: (ev: FocusEvent<HTMLInputElement>) => void;
  onBlur?: (ev: FocusEvent<HTMLInputElement>) => void;
}

export const DEFAULT_DEBOUNCE_INTERVAL_IN_MS = 300;

const BaseInput: FC<BaseInputProps> = props => {
  const {
    id,
    value,
    placeholder,
    pattern,
    disabled,
    debounce = false,
    onChange,
    onDebounceChange,
    onFocus,
    onBlur,
  } = props;
  const time = useRef<NodeJS.Timeout>();
  const valueToBeUpdated = useRef<string>();

  const debounceInterval: number = useMemo(() => {
    if (!debounce) {
      return 0;
    }
    if (debounce === true) {
      return DEFAULT_DEBOUNCE_INTERVAL_IN_MS;
    }
    return debounce as number;
  }, [debounce]);

  const handleChange = useCallback(
    (ev: ChangeEvent<HTMLInputElement>) => {
      if (!pattern || !ev.target.validity.patternMismatch) {
        valueToBeUpdated.current = ev.currentTarget.value;
        onChange?.(valueToBeUpdated.current);
        if (debounceInterval) {
          if (time.current) {
            clearTimeout(time.current);
          }
          time.current = setTimeout(() => {
            onDebounceChange?.(valueToBeUpdated.current as string);
            time.current = undefined;
            valueToBeUpdated.current = undefined;
          }, debounceInterval);
        } else {
          onDebounceChange?.(valueToBeUpdated.current);
        }
      }
    },
    [pattern, debounceInterval, onChange, onDebounceChange],
  );

  const handleBlur = useCallback(
    (ev: FocusEvent<HTMLInputElement>) => {
      if (time.current) {
        clearTimeout(time.current);
        time.current = undefined;
      }
      if (valueToBeUpdated.current !== undefined) {
        onDebounceChange?.(valueToBeUpdated.current);
      }
      onBlur?.(ev);
    },
    [onDebounceChange, onBlur],
  );

  return (
    <input
      className="tc__base-input"
      id={id}
      type="text"
      value={value}
      placeholder={placeholder}
      pattern={pattern}
      disabled={disabled}
      onChange={handleChange}
      onFocus={onFocus}
      onBlur={handleBlur}
    />
  );
};

export default BaseInput;
