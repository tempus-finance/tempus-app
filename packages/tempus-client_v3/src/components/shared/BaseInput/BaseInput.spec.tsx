import { act, fireEvent, render, waitFor } from '@testing-library/react';
import { FC, useState } from 'react';
import BaseInput, { BaseInputProps } from './BaseInput';

const mockOnChange = jest.fn();
const mockOnDebounceChange = jest.fn();
const mockOnBlur = jest.fn();
const mockOnFocus = jest.fn();

const defaultProps: BaseInputProps = {
  value: '',
  placeholder: 'base input placeholder',
  pattern: '',
  disabled: false,
  onChange: mockOnChange,
  onDebounceChange: mockOnDebounceChange,
  onBlur: mockOnBlur,
  onFocus: mockOnFocus,
};

const Wrapper: FC<BaseInputProps> = props => {
  const [value, setValue] = useState<string>(props.value ?? '');
  mockOnChange.mockImplementation((val: string) => setValue(val));
  return <BaseInput {...props} value={value} />;
};

const subject = (props: BaseInputProps) => render(<Wrapper {...props} />);

describe('input', () => {
  beforeEach(jest.useFakeTimers);
  afterEach(jest.useRealTimers);

  it('renders a base input with value', () => {
    const props = { ...defaultProps, value: 'value' };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();
    expect(input).toHaveValue(props.value);

    expect(input).toMatchSnapshot();
  });

  it('renders a base input with pattern', () => {
    const props = { ...defaultProps, pattern: '[0-9,]*[.]?[0-9]*' };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();
    expect(input).not.toBeNull();
    expect(input.getAttribute('pattern')).toBe(props.pattern);

    expect(input).toMatchSnapshot();
  });

  it('renders a disabled base input', () => {
    const props = { ...defaultProps, disabled: true };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();
    expect(input).not.toBeNull();
    expect(input.getAttribute('disabled')).not.toBeNull();

    expect(input).toMatchSnapshot();
  });

  it('type text in base input will trigger onChange', () => {
    const { getByRole } = subject(defaultProps);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.change(input, { target: { value: 'abcd' } });

    expect(mockOnChange).toHaveBeenCalledWith('abcd');
    expect(mockOnDebounceChange).toHaveBeenCalledWith('abcd');
    expect(input).toHaveValue('abcd');
  });

  it('focus in base input will trigger onFocus', () => {
    const { getByRole } = subject(defaultProps);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.focus(input);

    expect(mockOnFocus).toHaveBeenCalled();
  });

  it('focus in base input will trigger onBlur', () => {
    const { getByRole } = subject(defaultProps);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.blur(input);

    expect(mockOnBlur).toHaveBeenCalled();
  });

  it('type text with match pattern in base input will trigger onChange and onDebounceChange', () => {
    const props = { ...defaultProps, pattern: '[0-9,]*[.]?[0-9]*' };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.change(input, { target: { value: '1234' } });

    expect(mockOnChange).toHaveBeenCalledWith('1234');
    expect(mockOnDebounceChange).toHaveBeenCalledWith('1234');
    expect(input).toHaveValue('1234');
  });

  it('type text with unmatch pattern in base input will not trigger onChange and onDebounceChange', () => {
    const props = { ...defaultProps, pattern: '[0-9,]*[.]?[0-9]*' };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.change(input, { target: { value: 'abcd' } });

    expect(mockOnChange).not.toHaveBeenCalled();
    expect(mockOnDebounceChange).not.toHaveBeenCalled();
    expect(input).not.toHaveValue('abcd');
  });

  it('type text in base input with zero debounce will trigger onChange and onDebounceChange immediately', () => {
    const props = { ...defaultProps, debounce: 0 };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.change(input, { target: { value: 'abcd' } });

    expect(mockOnChange).toHaveBeenCalledWith('abcd');
    expect(mockOnDebounceChange).toHaveBeenCalledWith('abcd');
    expect(input).toHaveValue('abcd');
  });

  it('type text in base input with debounce will trigger onDebounceChange after 300ms', () => {
    const props = { ...defaultProps, debounce: true };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.change(input, { target: { value: 'abcd' } });

    expect(mockOnChange).toHaveBeenCalled();
    expect(mockOnDebounceChange).not.toHaveBeenCalled();
    expect(input).toHaveValue('abcd');

    act(() => {
      jest.advanceTimersByTime(300);
    });

    expect(mockOnDebounceChange).toHaveBeenCalledWith('abcd');
  });

  it('type text in base input with debounce 1000 will trigger onDebounceChange after 1000ms', () => {
    const props = { ...defaultProps, debounce: 1000 };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.change(input, { target: { value: 'abcd' } });

    expect(mockOnChange).toHaveBeenCalled();
    expect(mockOnDebounceChange).not.toHaveBeenCalled();
    expect(input).toHaveValue('abcd');

    act(() => {
      jest.advanceTimersByTime(300);
    });

    expect(mockOnDebounceChange).not.toHaveBeenCalled();

    act(() => {
      jest.advanceTimersByTime(700);
    });

    expect(mockOnDebounceChange).toHaveBeenCalledWith('abcd');
  });

  it('keep typing text in base input with debounce will trigger onDebounceChange after 300ms from stop typing', () => {
    const props = { ...defaultProps, debounce: true };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.change(input, { target: { value: 'a' } });

    expect(mockOnChange).toHaveBeenCalledWith('a');
    expect(mockOnDebounceChange).not.toHaveBeenCalled();
    expect(input).toHaveValue('a');

    act(() => {
      jest.advanceTimersByTime(200);
    });

    fireEvent.change(input, { target: { value: 'ab' } });

    expect(mockOnChange).toHaveBeenCalledWith('ab');
    expect(mockOnDebounceChange).not.toHaveBeenCalled();
    expect(input).toHaveValue('ab');

    act(() => {
      jest.advanceTimersByTime(200);
    });

    fireEvent.change(input, { target: { value: 'abc' } });

    expect(mockOnChange).toHaveBeenCalledWith('abc');
    expect(mockOnDebounceChange).not.toHaveBeenCalled();
    expect(input).toHaveValue('abc');

    act(() => {
      jest.advanceTimersByTime(200);
    });

    fireEvent.change(input, { target: { value: 'abcd' } });

    expect(mockOnChange).toHaveBeenCalledWith('abcd');
    expect(mockOnDebounceChange).not.toHaveBeenCalled();
    expect(input).toHaveValue('abcd');

    act(() => {
      jest.advanceTimersByTime(300);
    });

    expect(mockOnDebounceChange).toHaveBeenCalledWith('abcd');
  });

  it('no type but just blur in base input with debounce will not trigger onChange or onDebounceChange', () => {
    const props = { ...defaultProps, value: 'abcd', debounce: true };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.blur(input);

    expect(mockOnChange).not.toHaveBeenCalled();
    expect(mockOnDebounceChange).not.toHaveBeenCalled();
  });

  it('type text and then blur in base input with debounce will trigger onChange and onDebounceChange immediately', () => {
    const props = { ...defaultProps, value: 'ab', debounce: true };
    const { getByRole } = subject(props);

    const input = getByRole('textbox');

    expect(input).not.toBeNull();

    fireEvent.change(input, { target: { value: 'abcd' } });
    act(() => {
      jest.advanceTimersByTime(100);
    });
    fireEvent.blur(input);

    expect(mockOnChange).toHaveBeenCalledWith('abcd');
    expect(mockOnDebounceChange).toHaveBeenCalledWith('abcd');
    expect(input).toHaveValue('abcd');
  });
});
