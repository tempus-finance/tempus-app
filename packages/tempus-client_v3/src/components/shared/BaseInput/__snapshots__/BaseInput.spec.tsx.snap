// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input renders a base input with pattern 1`] = `
<input
  class="tc__base-input"
  pattern="[0-9,]*[.]?[0-9]*"
  placeholder="base input placeholder"
  type="text"
  value=""
/>
`;

exports[`input renders a base input with value 1`] = `
<input
  class="tc__base-input"
  pattern=""
  placeholder="base input placeholder"
  type="text"
  value="value"
/>
`;

exports[`input renders a disabled base input 1`] = `
<input
  class="tc__base-input"
  disabled=""
  pattern=""
  placeholder="base input placeholder"
  type="text"
  value=""
/>
`;
